src/nn/add_into.rs

```diff
@@ -23,26 +23,20 @@ use super::*;
 #[derive(Debug, Default, Clone)]
 pub struct AddInto<T>(pub T);
 
-impl<T: BuildOnDevice<D, E>, D: DeviceStorage, E: Dtype> BuildOnDevice<D, E> for AddInto<T> {
+impl<T: BuildOnDevice<D, E>, D: Device<E>, E: Dtype> BuildOnDevice<D, E> for AddInto<T> {
     type Built = AddInto<T::Built>;
 }
 
-impl<T: BuildModule<D, E>, D: DeviceStorage, E: Dtype> BuildModule<D, E> for AddInto<T> {
-    fn try_build(device: &D) -> Result<Self, <D>::Err> {
-        Ok(Self(BuildModule::try_build(device)?))
-    }
-}
-
-impl<E: Dtype, D: DeviceStorage, T: TensorCollection<E, D>> TensorCollection<E, D> for AddInto<T> {
-    fn iter_tensors<V: ModuleVisitor<Self, E, D>>(visitor: &mut V) -> Result<(), V::Err> {
-        visitor.visit_module("0", |s| &s.0, |s| &mut s.0)
-    }
-}
+impl<E: Dtype, D: Device<E>, T: TensorCollection<E, D>> TensorCollection<E, D> for AddInto<T> {
+    type Output<E2: Dtype, D2: Device<E2>> = AddInto<T::Output<E2, D2>>;
 
-impl<T: ToDevice<D>, D> ToDevice<D> for AddInto<T> {
-    type Output = AddInto<T::Output>;
-    fn to_device(&self, device: &D) -> Self::Output {
-        AddInto(self.0.to_device(device))
+    fn iter_tensors<E2: Dtype, D2: Device<E2>, V: ModuleVisitor<Self, E, D, E2, D2>>(
+        visitor: &mut V,
+    ) -> Result<Option<Self::Output<E2, D2>>, V::Err> {
+        visitor.visit_fields(
+            ModuleField::new("0", |s: &Self| &s.0, |s| &mut s.0),
```

Comment:
I think we can get around having to specify `&Self` on these by adding a default method in the TensorCollection trait:

```rust
trait TensorCollection {
   fn iter_tensors(...);
    fn module(...) -> ModuleField { ... }
    fn tensor(...) -> TensorField { ... }
}
```

and then we could call like:
```suggestion
            Self::module("0", |s| &s.0, |s| &mut s.0),
```